server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -99,9 +113,43 @@ synchronized void add(final long currentGlobalCheckpoint, final GlobalCheckpoint
             executor.execute(() -> notifyListener(listener, lastKnownGlobalCheckpoint, null));
         } else {
             if (listeners == null) {
-                listeners = new ArrayList<>();
+                listeners = new LinkedHashMap<>();
+            }
+            if (timeout == null) {
+                listeners.put(listener, null);
+            } else {
+                listeners.put(
+                        listener,
+                        scheduler.schedule(
+                                () -> {
+                                    final boolean removed;
+                                    synchronized (this) {
+                                        if (listeners == null) {
```

Comment:
how about `removed = listeners != null && listeners.remove(listener) != null` ?

Reply:
So I initially opted for not this approach because it requires a little extra thought to realize it's correct. Namely the fact that we put null for listeners that do not have a timeout is not necessarily an indication that listeners.remove(listener) being null means the listener was not in the map. However, it does mean that for this section of code since this section of code only runs if a timeout fires, which means the listener had to have at some point a timeout associated with it. So because this approach requires a little extra thought I opted for the approach I wrote in the version of this pull request. However, your push forced me to reconsider and I reworded the comments a little bit to help a future reader of this code see that little extra thought. I pushed d9abe55e9ee33386310b1136e7ef49a868c5dbc7.